core/backend/src/test/AdvancedEqual.ts

```diff
@@ -0,0 +1,132 @@
+  considerNonExistingAndUndefinedEqual?: boolean;
+  /** Specify standard property keys holding ec class identifiers (when true, `classFullName` and `relClassName`)
+   * to be normalized during comparison so that differently cased and alternative syntax still is considered equal
+   * Passing `true`, is the same as passing `["classFullName", "relClassName"]`
+   */
+  normalizeClassNameProps?: boolean | string[];
+}
+
+export const defaultOpts = {
+  tolerance: 1e-10,
+  considerNonExistingAndUndefinedEqual: false,
+} as const;
+
+declare global {
+  namespace Chai {
+    interface Deep {
+      advancedEqual(actual: any, options?: DeepEqualOpts): Assertion;
+      subsetEqual(actual: any, options?: DeepEqualOpts): Assertion;
+    }
+  }
+}
+
+/** get whether two numbers are almost equal within a tolerance  */
+const isAlmostEqualNumber: (a: number, b: number, tol: number) => boolean = Geometry.isSameCoordinate;
+
+/** normalize a classname for comparisons */
+const normalizeClassName = (name: string) => name.toLowerCase().replace(/:/, ".");
+
+interface AdvancedEqualFuncOpts extends DeepEqualOpts {
+  /** only test */
+  useSubsetEquality?: boolean;
+}
+
+/**
+ * The diff shown on failure will show undefined fields as part of the diff even if
+ * consideringNonExistingAndUndefinedEqual is true. You can ignore that.
+ * We would have to mutate the object to clean the diff, although it is worth considering.
+ */
+export function advancedDeepEqual(
+  e: any,
+  a: any,
+  options: AdvancedEqualFuncOpts = {},
+): boolean {
+  const normalizedClassNameProps
+    = options.normalizeClassNameProps === true
+      ? ["classFullName", "relClassName"]
+      : options.normalizeClassNameProps || [];
+  if (options.tolerance === undefined) options.tolerance = defaultOpts.tolerance;
+  if (e === a) return true;
+  if (typeof e !== typeof a) return false;
+  switch (typeof e) {
+    case "number":
+      return isAlmostEqualNumber(e, a, options.tolerance);
+    case "string":
+    case "boolean":
+    case "function":
+    case "symbol":
+    case "undefined":
+      return false; // these objects can only be strict equal which was already tested
+    case "object":
+      if ((e === null) !== (a === null)) return false;
+      const eSize = Object.keys(e).filter((k) => options.considerNonExistingAndUndefinedEqual && e[k] !== undefined).length;
+      const aSize = Object.keys(a).filter((k) => options.considerNonExistingAndUndefinedEqual && a[k] !== undefined).length;
+      return (eSize === aSize || !!options.useSubsetEquality) && Object.keys(e).every(
+        (keyOfE) =>
+          (keyOfE in a || options.considerNonExistingAndUndefinedEqual) &&
+          normalizedClassNameProps.includes(keyOfE)
+            ? advancedDeepEqual(normalizeClassName(e[keyOfE]), normalizeClassName(a[keyOfE]))
+            : advancedDeepEqual(e[keyOfE], a[keyOfE], options)
+      );
+    default: // bigint unhandled
+      throw Error("unhandled deep compare type");
```

Comment:
consider including type name in error for more context

Reply:
done

Reply:
done